aesd-char-driver/main.c

```diff
@@ -0,0 +1,304 @@
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev* device;
+
+    PDEBUG("open");
+    /**
+     * TODO: handle open
+     */
+
+    // Getting the pointer of the sturcure that containes the cdev
+    device = container_of(inode->i_cdev, struct aesd_dev, cdev);
+
+    // Storing the device pointer in the private data for other modules to use it
+    filp->private_data = device;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *device = filp->private_data;
+    struct aesd_buffer_entry *temp_kernel_buff_ptr;
+    size_t offset_byte_loc;
+    int temp_count = 0;
+
+    PDEBUG("Reads %zu bytes with offset %lld", count, *f_pos);
+    /**
+     * TODO: handle read
+     */
+
+    // Obtaining the lock 
+    mutex_lock(&aesd_device.lock);
+
+    // Getting the buffer and the position of the offset
+    temp_kernel_buff_ptr = aesd_circular_buffer_find_entry_offset_for_fpos(&device->aesd_circular_buffer,
+                                                                           *f_pos, 
+                                                                           &offset_byte_loc);
+    if(temp_kernel_buff_ptr == NULL)
+    {
+        *f_pos = 0;
+        goto out;
+    }
+
+    // 
+    if ((temp_kernel_buff_ptr->size - offset_byte_loc) < count) 
+    {
+        *f_pos += (temp_kernel_buff_ptr->size - offset_byte_loc);
+        temp_count = temp_kernel_buff_ptr->size - offset_byte_loc;
+    } 
+    else 
+    {
+        *f_pos += count;
+        temp_count = count;
+    }
+
+    if (copy_to_user(buf, temp_kernel_buff_ptr->buffptr+offset_byte_loc, temp_count)) 
+    {
+        retval = -EFAULT;
+        goto out;
+    }
+
+    retval = temp_count;
+
+    // Releasing the lock 
+    out: mutex_unlock(&aesd_device.lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = count;
+    char* temp_kernel_buff_ptr = NULL;
+    bool newline_recieved = false;
+    int status;
+    int i;
+    char* ret_buff_ptr = NULL;
+    struct aesd_buffer_entry aesd_buffer_write_entry;
+    char* temp = NULL;
+
+    
+    // Obtaining the device pointer ans storing in a local variable for our use
+    struct aesd_dev* device = filp->private_data;
+    /**
+     * TODO: handle write
+     */
+
+    // Debug prints
+    PDEBUG("Writing %zu bytes with offset %lld", count, *f_pos);
+    
+    // Obtaining the lock
+    mutex_lock(&aesd_device.lock);
+
+    // Mallocing user space buffer into a temp kernel space buffer
+    temp_kernel_buff_ptr = (char*)kmalloc(count, GFP_KERNEL);
```

Comment:
@nihalthirunakarasu does this block and the one below need to be locked?  What is `aesd_device.lock` locking?